Sources/SwiftParser/Patterns.swift

```diff
@@ -194,11 +198,12 @@ extension Parser.Lookahead {
     }
 
     if !self.at(.rightParen) {
+      var loopProgress = LoopProgressCondition()
       repeat {
         guard self.canParsePattern() else {
           return false
         }
-      } while self.consume(if: .comma) != nil
+      } while loopProgress.evaluate(currentToken) && self.consume(if: .comma) != nil
     }
```

Comment:
In lookahead, we should consume the token, then check for progress.

Reply:
I think we would hit the loop progress assertion in either case but if we move the check after `consume` we might be able to abort the loop one iteration earlier. 

For consistency I moved `loopProgress.evaluate` to be the last condition of all loops. On second thought that also reads nicer because it makes the actual condition more prominent.